{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## answers 02: the adventure of the ten Arcs\n",
    "**Sean's answers**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. install kallisto\n",
    "\n",
    "`kallisto` is an example of the sort of code that I'm interested not only in using (as a program), but studying (as code), so I downloaded its source code and built my own binaries. When I install bioinformatics software from source code, I keep it in a directory of mine called `~/alien-src`. \n",
    "\n",
    "I work on a Mac OS/X laptop, and I prefer working on a unix command line, so my steps will reflect that.\n",
    "\n",
    "Following the instructions at [the kallisto site](https://pachterlab.github.io/kallisto/source), first I installed some dependencies. These are examples of things that I don't currently care what's in their source code; I just want them to work, so I'm very happy to be able to install them with a good package manager. To install the dependencies, I installed [homebrew](http://brew.sh) first.\n",
    "\n",
    "```bash\n",
    "# install homebrew\n",
    "% /usr/bin/ruby -e \"$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)\"\n",
    "  \n",
    "# install cmake\n",
    "% brew install cmake\n",
    "   \n",
    "# install the HDF5 library\n",
    "% brew install hdf5\n",
    "```\n",
    "\n",
    "Now for the kallisto code:\n",
    "\n",
    "```bash\n",
    "% wget https://github.com/pachterlab/kallisto/archive/v0.46.1.tar.gz\n",
    "% tar xzf v0.46.1.tar.gz\n",
    "% cd kallisto-0.46.1/\n",
    "% cd ext/htslib\n",
    "% autoheader\n",
    "% autoconf\n",
    "% cd ../..\n",
    "\n",
    "% mkdir build\n",
    "% cd build\n",
    "% cmake ..\n",
    "% make\n",
    "% make install\n",
    "```\n",
    "\n",
    "Now I have `kallisto` in `/usr/local/bin/kallisto`. If I run it without arguments, it's supposed to give me some basic help info..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kallisto 0.46.1\r\n",
      "\r\n",
      "Usage: kallisto <CMD> [arguments] ..\r\n",
      "\r\n",
      "Where <CMD> can be one of:\r\n",
      "\r\n",
      "    index         Builds a kallisto index \r\n",
      "    quant         Runs the quantification algorithm \r\n",
      "    bus           Generate BUS files for single-cell data \r\n",
      "    pseudo        Runs the pseudoalignment step \r\n",
      "    merge         Merges several batch runs \r\n",
      "    h5dump        Converts HDF5-formatted results to plaintext\r\n",
      "    inspect       Inspects and gives information about an index\r\n",
      "    version       Prints version information\r\n",
      "    cite          Prints citation information\r\n",
      "\r\n",
      "Running kallisto <CMD> without arguments prints usage information for <CMD>\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "# Do we have kallisto installed? If we do, `kallisto` alone shows help info.\n",
    "! kallisto"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Reproduce Moriarty's result\n",
    "\n",
    "I have Moriarty's two data files [arc.fasta.gz](arc.fasta.gz) and [arc.fastq.gz](arc.fastq.gz) in my current working directory where I'm running my notebook page. \n",
    "\n",
    "Within the notebook page, we can run command line programs and create output files.\n",
    "\n",
    "In the `kallisto quant` command, I have to tell `kallisto` that these are single ended data, with a fragment length distribution of mean 150, sd 20, as defined by the pset: these are the `--single -l 150 -s 20` options.\n",
    "\n",
    "I'm sending the output into the current working directory, with `-o .`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[build] loading fasta file arc.fasta.gz\n",
      "[build] k-mer length: 31\n",
      "[build] counting k-mers ... done.\n",
      "[build] building target de Bruijn graph ...  done \n",
      "[build] creating equivalence classes ...  done\n",
      "[build] target de Bruijn graph has 19 contigs and contains 10000 k-mers \n",
      "\n",
      "\n",
      "[quant] fragment length distribution is truncated gaussian with mean = 150, sd = 20\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 10\n",
      "[index] number of k-mers: 10,000\n",
      "[index] number of equivalence classes: 26\n",
      "[quant] running in single-end mode\n",
      "[quant] will process file 1: arc.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 100,000 reads, 99,981 reads pseudoaligned\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 52 rounds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create the kallisto index of the Arc transcriptome\n",
    "! kallisto index -i arc.idx arc.fasta.gz\n",
    "\n",
    "# Analyze the reads\n",
    "! kallisto quant -i arc.idx -o . --single -l 150 -s 20 arc.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now there's three new files in the directory. `abundance.h5` is an HDF5 file that we're not going to use. `arc.idx` is the kallisto index of the transcriptome. `abundance.tsv` is what we care about right now: the quantitation results. We still don't have to use python - we can just `cat` the file from the command line to look at it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 19700\r\n",
      "-rw-r--r-- 1 seddy staff    35848 Sep 25 08:40 abundance.h5\r\n",
      "-rw-r--r-- 1 seddy staff      352 Sep 25 08:40 abundance.tsv\r\n",
      "-rw-r--r-- 1 seddy staff      329 Sep 25 08:40 run_info.json\r\n"
     ]
    }
   ],
   "source": [
    "! ls -lt | head -4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target_id\tlength\teff_length\test_counts\ttpm\r\n",
      "Arc1\t4000\t3851\t3348.79\t24514.5\r\n",
      "Arc2\t2000\t1851\t3574.56\t54440.7\r\n",
      "Arc3\t3000\t2851\t28065.2\t277510\r\n",
      "Arc4\t4000\t3851\t10597.7\t77579.1\r\n",
      "Arc5\t4000\t3851\t12526.7\t91700.1\r\n",
      "Arc6\t3000\t2851\t1953.39\t19315.2\r\n",
      "Arc7\t2000\t1851\t5579.78\t84980.4\r\n",
      "Arc8\t2000\t1851\t5700.81\t86823.7\r\n",
      "Arc9\t3000\t2851\t3052.55\t30183.8\r\n",
      "Arc10\t3000\t2851\t25581.6\t252953\r\n"
     ]
    }
   ],
   "source": [
    "! cat abundance.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yup, those are Moriarty's results. He reported these TPM calls, rounded to two sig figs. (That Moriarty, he does know how to round off numbers properly, gotta give him that.)\n",
    "\n",
    "So we've reproduced his result, and now we can think about why it wasn't what we expected. \n",
    "\n",
    "(From the perspective of the pset: this step made sure that you've got kallisto installed and you know how to run it!)\n",
    "\n",
    "Let's pull in those numbers for later use, and we'll do it with a little function that we're reuse later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_tpms(tsvfile):\n",
    "    i = 0\n",
    "    tpm = []\n",
    "    with open(tsvfile) as f:\n",
    "        f.readline() # skip header\n",
    "        for line in f:\n",
    "            tpm.append(float(line.split()[4]))\n",
    "    return tpm\n",
    "\n",
    "tpm_moriarty = get_tpms('abundance.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## 3. simulate an Arc transcriptome and RNA-seq reads\n",
    "\n",
    "This is a small-ish example of **doing controls with synthetic data**. How much should we trust Moriarty's result with `kallisto`? With any new computational method we're trying out, it can be worthwhile to see how it performs on synthetic data where we know the answer. We know the parameters we used to generate the data, and we can see if the method can infer them.\n",
    "\n",
    "You might say, just because the method works on synthetic data doesn't mean that it works on real data, because real data are going to be more complex. You'd be right. But the point is, if a method has problems getting the right answer on an _easy_ problem, it's not likely to do better on _harder_ ones. This is not much different than doing experimental controls, where your positive control is often an easier experiment with well-defined, well-known reagents and inputs. Just because your positive control worked, doesn't mean your experiment worked; but if your positive control failed, you should worry.\n",
    "\n",
    "Specifically, we're going to simulate Arc read data using _our_ RNA abundance parameters. If `kallisto` infers those parameters successfully, then we still don't know where the discrepancy between our results and Moriarty's is. But if `kallisto` infers _Moriarty's_ result when we give it synthetic data that have been explicitly generated using _our_ abundance parameters, then the problem lies in `kallisto`, or in how we're using it.\n",
    "\n",
    "I'm going to cheat a little here: I'm going to go ahead and write in code that I need for part 5 of the answer, and when I get there I'm going to rerun the notebook page with different parameter settings. A more elegant strategy would be to modularize my simulator inside a Python function that I can call with different options.\n",
    "\n",
    "I'm going to show the more difficult (3b) path through the pset here, which includes sampling a fragment length, simulating reads in both orientations, and simulating occasional base-calling errors in the reads. I'll put comments on stuff that you would just skip over in the easier (3a) path through the pset.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy  as np\n",
    "import string\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Options that control how I'm going to run the simulation\n",
    "do_pset       = True    # Use exactly the Arc structure and abundance params from the pset\n",
    "do_no_overlap = False   # Variant with all L[i] = one segment so there's no overlap (makes inference problem trivial)\n",
    "do_no_circle  = False   # Variant with no circular permutation (as we'll see, one candidate/suspected problem in kallisto)\n",
    "\n",
    "# Names of files we're going to generate: simdata.{fasta,fastq,dat}\n",
    "outpfx     = 'simdata'\n",
    "fasta_file = '{}.fasta'.format(outpfx)  # transcriptome file\n",
    "fastq_file = '{}.fastq'.format(outpfx)  # reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Set up the Arc locus \n",
    "#\n",
    "S         = 10                     # Number of segments in the Arc locus (A..J)\n",
    "T         = S                      # Number of different transcripts (the same, one starting on each segment, 1..10)\n",
    "N         = 100000                 # total number of observed reads we will generate \n",
    "alpha     = 0.999                  # base calling accuracy (Q30 bases, typical of Illumina) ... unused in (3a) version\n",
    "len_S     = 1000                   # length of each segment (nucleotides)\n",
    "len_Arc   = len_S * S              # total length of the Arc locus (nucleotides)\n",
    "len_R     = 75                     # read length\n",
    "mean_frag = 150                    # fragment size: mean (of a truncated Gaussian)  ... unused in (3a) version\n",
    "sd_frag   = 20                     # fragment size: stdev                           ... unused in (3a) version\n",
    "\n",
    "Slabel  = list(string.ascii_uppercase)[:S]               # ['A'..'J']        : the upper case labels for Arc locus segments \n",
    "Tlabel  = [ \"Arc{}\".format(d) for d in range(1,T+1) ]    # ['Arc1'..'Arc10'] : the labels for Arc transcript isoforms\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Encode base accuracy, Sanger format: we only need one QV for all our data, keep that character code in <qvchar>\n",
    "# kallisto ignores this information, so we could just as easily just set this to any valid QV, like '?' (Q30) or 'I' (Q40),\n",
    "# which is what you'd do in version (3a) where you wouldn't be using <alpha> and introducing simulated basecall errors.\n",
    "qv     = int(round(-10 * np.log10 ( 1. - alpha)))\n",
    "if qv > 93: qv = 93\n",
    "if qv < 0:  qv = 0\n",
    "qvchar = chr(qv+33)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Set up the parts of the Arc locus model that we can randomize, if we want.\n",
    "\n",
    "# The do_pset option sets up exactly the same parameters as our params from the pset\n",
    "# We can also get the same result by setting the numpy RNG seed to 20160925, \n",
    "# if you don't alter the order in which we're calling np.random.\n",
    "#\n",
    "if do_pset:\n",
    "    nu = np.array([0.008, 0.039, 0.291, 0.112, 0.127, 0.008, 0.059, 0.060, 0.022, 0.273])\n",
    "    L  = np.array([4000,  2000,  3000,  4000,  4000,  3000,  2000,  2000,  3000,  3000]).astype(int)\n",
    "else:\n",
    "    nu = np.random.dirichlet(np.ones(T))                       # true expression levels \\nu: P(Ti | nu_i)\n",
    "    L  = [ len_S * np.random.randint(2,5) for _ in range(T) ]  # each isoform i is 2..4 segments long\n",
    "# (what's np.random.dirichlet() do, you ask? It's a useful trick for sampling a uniformly-distributed probability vector, using a so-called Dirichlet distribution.)\n",
    "\n",
    "\n",
    "# The do_no_overlap option forces each transcript to be 1 segment long, thus nonoverlapping.     \n",
    "if do_no_overlap:\n",
    "    L    =  np.multiply(np.ones(T), len_S).astype(int)            \n",
    "    \n",
    "# The do_no_circle option truncates lengths L[i] as needed to avoid any overlap due to circular permutation\n",
    "if do_no_circle:   \n",
    "    for i in range(T):\n",
    "        if i*1000 + L[i] > len_Arc:\n",
    "            L[i] = len_Arc - i*1000    # i.e. the Arc that starts at 9 must be 1 long. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Make extra sure that the theta's sum to one. Due to roundoff error, the pset numbers don't quite!\n",
    "Z  = np.sum(nu)\n",
    "nu = np.divide(nu, Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**We generate a random Arc DNA sequence for the whole locus:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample the Arc locus nucleotide sequence, as a string.\n",
    "#\n",
    "Arc_locus = ''.join(np.random.choice(list('ACGT'), len_Arc))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**then we extract the individual Arc transcript sequences:**\n",
    "\n",
    "We have to deal with Arc being a circle. When we go to extract an Arc transcript that rolls around the circle, we have to do something to extract it in two pieces from the linear `Arc_locus[]` list we just generated.\n",
    "\n",
    "There's other ways to do this too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# From that, extract the individual Arc1-Arc10 transcript sequences:\n",
    "#\n",
    "Arc = []\n",
    "for i in range(T):\n",
    "    start = i * len_S\n",
    "    end   = (start + L[i] - 1) % len_Arc   # circularly permuted!\n",
    "    if end < start:\n",
    "        Arc.append( Arc_locus[start:] + Arc_locus[:end+1] )  # watch out for that circular permutation; fetch in two pieces.\n",
    "    else:\n",
    "        Arc.append( Arc_locus[start:end+1] )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**write the transcriptome file, in FASTA format:** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write the Arc1-Arc10 transcriptome to a FASTA file.\n",
    "#\n",
    "def write_fasta(f, seqname, seq):\n",
    "    print(\">{}\".format(seqname), file=f)\n",
    "    for i in range(0, len(seq), 60):\n",
    "        if i+60 > len(seq):\n",
    "            line = ''.join(seq[i:])\n",
    "        else:\n",
    "            line = ''.join(seq[i:i+60])\n",
    "        print(line, file=f)\n",
    "\n",
    "with open(fasta_file, 'w') as f:\n",
    "    for i in range(T):\n",
    "        write_fasta(f, Tlabel[i], Arc[i])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "that file should be there now, right?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Arc1\r\n",
      "CAGTACTATTGACAAAAAGGGTCAATCCGTCGTTGCAAGGCGGTCGATAGTTAAGCCTAT\r\n",
      "TTTCTTTGCGCGTAGAAGTGTGGCTCGAAGGTTCCCATTGGGCTAACGCGCCGGGTAACA\r\n",
      "TGGCCGTCTGCATCCTTAGGTAGCCCGGAAGGGTCTAAACCATGTCACGCGTTCCTAGGT\r\n",
      "AGTTGGTGACTCACGCGGTAATTAGAGCTTCAGACCACTTCCTTACCAGTATCTGCCATA\r\n",
      "TCCCCGAAATCGGCCATGAAACGCGTAACTATAGAGTTCTGTCTAGGAGAGGCGGATCAG\r\n",
      "TCCCACGGGAACAGTCGACGATCACGGCCGCCAGGTGTGGGTGGGGGGAGACAAAAAATG\r\n",
      "CAAGATTATATTTGTCACGATTGCTACGCTATGGATCAGACTAAGATTACAAACTTCACG\r\n",
      "GCCACAGCCAAAAACAAATACTTAACTCTGAAACGCCTGAGCCCTGGACACTTTCCGCCT\r\n",
      "CTCCCGGACCGCGTTTACCCATTAATACACGGGCTCGCTAGAGCCTGGAACCGATTGCGC\r\n"
     ]
    }
   ],
   "source": [
    "! head simdata.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "right."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**generate reads, and write them to their output FASTQ file as we go:**\n",
    "\n",
    "(Takes a bit of time. Watch the `In [*]` to the left; while it's still an asterisk, it's still running the code.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Generate reads\n",
    "#\n",
    "with open(fastq_file, 'w') as f:\n",
    "    for n in range(N):                         # for each independent read that we sample:\n",
    "\n",
    "        i    = np.random.choice(T, p=nu)       # choice of isoform i = 0..T-1, according to nucleotide abundance nu_i       \n",
    "\n",
    "        while True:                            # sample a random fragment length from a truncated Gaussian, >= readlength, <= L. (unused in (3a) version)\n",
    "            flen = int(np.random.normal(mean_frag, sd_frag))\n",
    "            if flen >= len_R: break\n",
    "        if flen > L[i]: flen = L[i]            # in the simpler (3a) version you could just set flen=75 ... then remember to set your `kallisto quant` frag len \n",
    "                                               # parameters accordingly, with frag len of 75 and a small nonzero std dev, something like `-l 75 -s 5`\n",
    "\n",
    "        j    = np.random.choice(L[i]-flen+1)   # choice of start position j = 0..L[i]-flen\n",
    "        if np.random.choice(2) == 0:           # orientation: 0 = sense, 1 = antisense  ... in the simpler (3a) version you'd just sample the sense orientation.\n",
    "            read = Arc[i][j:j+len_R]           # extract the read sequence \n",
    "        else:        \n",
    "            read = Arc[i][j+flen-1:j+flen-len_R-1:-1]             # fetching it in reverse orientation! note the 3rd arg, -1\n",
    "            read = read.translate(str.maketrans('ACGT', 'TGCA'))  # one of many ways you can complement a sequence\n",
    "\n",
    "        # Now dirty it up: add random basecall errors, according to accuracy <alpha>\n",
    "        # You'd skip this in the (3a) version.\n",
    "        seq = list(read)\n",
    "        for i in range(len(seq)):\n",
    "            if np.random.uniform() > alpha:\n",
    "                c = seq[i]\n",
    "                while c == seq[i]:                      # this `while` loop uniformly samples one of 3 possible mutations,\n",
    "                    c = np.random.choice(list('ACGT'))  # for any starting residue <seq[i]>\n",
    "                seq[i] = c\n",
    "        read = ''.join(seq)\n",
    "\n",
    "        print(\"@read{}\".format(n),        file=f)\n",
    "        print(read, file=f)\n",
    "        print(\"+\",  file=f)\n",
    "        print(\"{}\".format(''.join(qvchar*len_R)), file=f)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we've got a read file now, right?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@read0\r\n",
      "GCCCGCTTGTATGATCGATCACGTTCTTGGTGTTCCATCTTTGGATGACTGAACCGCGTCTTTCATGGTTAGGTA\r\n",
      "+\r\n",
      "???????????????????????????????????????????????????????????????????????????\r\n",
      "@read1\r\n",
      "GTGCTGATTTACACGCGTCCCGTTGTATTGTGTTCGCCTTAGTCTTAATGCTTCGACTCCCAGTCTATGTTAGGG\r\n",
      "+\r\n",
      "???????????????????????????????????????????????????????????????????????????\r\n",
      "@read2\r\n",
      "CTGTCCAAAGACTCGATCGGGAGTTGTGCTCTATACCATGCCGGCCATATATCACGCTCGGCGGAGAGTTATCAC\r\n"
     ]
    }
   ],
   "source": [
    "! head simdata.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "right. (The question marks look scary, but that's the QV code for a Q30 base.)\n",
    "\n",
    "Now let's deal with a pesky detail before we output the final summary. Our $\\nu_i$ are _nucleotide abundances_; to compare to the `kallisto` TPM output, we convert to per-transcript abundances in TPM. Now `tpm[]` are our known abundances in TPM, which we'll compare to `kallisto`'s inferred TPM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# tpm versus theta, per-transcript vs per-nucleotide abundance\n",
    "#\n",
    "tpm = np.array( [ nu[i] / L[i] for i in range(T) ])\n",
    "Z   = np.sum(tpm)\n",
    "tpm = np.divide(tpm, Z)\n",
    "tpm = np.multiply(tpm, 1000000)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**print info about the simulation, including (most importantly) the true Arc abundances:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The 10 transcripts of the sand mouse Arc locus:\n",
      "| transcript | nu    | TPM      | length | segments_covered |\n",
      "| -----------|-------|----------|------- | ---------------- |\n",
      "| Arc1       | 0.008 |   5904.1 |   4000 | ABCD\n",
      "| Arc2       | 0.039 |  57564.6 |   2000 | BC\n",
      "| Arc3       | 0.291 | 286346.9 |   3000 | CDE\n",
      "| Arc4       | 0.112 |  82656.8 |   4000 | DEFG\n",
      "| Arc5       | 0.127 |  93726.9 |   4000 | EFGH\n",
      "| Arc6       | 0.008 |   7872.1 |   3000 | FGH\n",
      "| Arc7       | 0.059 |  87084.9 |   2000 | GH\n",
      "| Arc8       | 0.060 |  88560.9 |   2000 | HI\n",
      "| Arc9       | 0.022 |  21648.2 |   3000 | IJA\n",
      "| Arc10      | 0.273 | 268634.7 |   3000 | JAB\n",
      "\n",
      "alpha:                  0.999\n",
      "mean fragment size:     150\n",
      "fragment size std dev.: 20\n",
      "read length:            75\n",
      "number of reads:        100000\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Output information on the transcript structure and expression level\n",
    "#\n",
    "print(\"The {} transcripts of the sand mouse Arc locus:\".format(T))\n",
    "print(\"| {0:10s} | {1:5s} | {2:8s} | {3:6s} | {4} |\".format(\"transcript\", \"nu\", \"TPM\", \"length\", \"segments_covered\"))\n",
    "print(\"| {0:10s}-|-{1:5s}-|-{2:8s}-|-{3:6s} | {4} |\".format('-'*10, '-'*5, '-'*8, '-'*6, '-'*16))\n",
    "for i in range(T):\n",
    "    print (\"| {0:10s} | {1:5.3f} | {2:8.1f} | {3:6d} | \".format(Tlabel[i], nu[i], tpm[i], L[i]), end=\"\")\n",
    "    for j in range(i,i+(L[i]//len_S)):\n",
    "        print(\"{}\".format(Slabel[j%S]), end=\"\")\n",
    "    print(\"\")\n",
    "print(\"\")\n",
    "\n",
    "print(\"alpha:                  {}\".format(alpha))\n",
    "print(\"mean fragment size:     {}\".format(mean_frag))\n",
    "print(\"fragment size std dev.: {}\".format(sd_frag))\n",
    "print(\"read length:            {}\".format(len_R))\n",
    "print(\"number of reads:        {}\".format(N))\n",
    "print(\"\")\n",
    "\n",
    "if do_no_circle:\n",
    "    print(\"(no circle variant)\")\n",
    "if do_no_overlap:\n",
    "    print(\"(no overlap variant)\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. test kallisto\n",
    "\n",
    "We're ready to run `kallisto` on the simulated data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[build] loading fasta file simdata.fasta\n",
      "[build] k-mer length: 31\n",
      "[build] counting k-mers ... done.\n",
      "[build] building target de Bruijn graph ...  done \n",
      "[build] creating equivalence classes ...  done\n",
      "[build] target de Bruijn graph has 19 contigs and contains 10000 k-mers \n",
      "\n",
      "\n",
      "[quant] fragment length distribution is truncated gaussian with mean = 150, sd = 20\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 10\n",
      "[index] number of k-mers: 10,000\n",
      "[index] number of equivalence classes: 26\n",
      "[quant] running in single-end mode\n",
      "[quant] will process file 1: simdata.fastq\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 100,000 reads, 99,989 reads pseudoaligned\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 55 rounds\n",
      "\n",
      "target_id\tlength\teff_length\test_counts\ttpm\n",
      "Arc1\t4000\t3851\t3167.09\t23193.2\n",
      "Arc2\t2000\t1851\t3371.18\t51362.8\n",
      "Arc3\t3000\t2851\t28453.9\t281461\n",
      "Arc4\t4000\t3851\t10692.4\t78302.6\n",
      "Arc5\t4000\t3851\t12499.6\t91537.1\n",
      "Arc6\t3000\t2851\t1822.16\t18024.4\n",
      "Arc7\t2000\t1851\t5646.27\t86025.7\n",
      "Arc8\t2000\t1851\t5696.99\t86798.4\n",
      "Arc9\t3000\t2851\t2785.11\t27549.8\n",
      "Arc10\t3000\t2851\t25854.2\t255745\n"
     ]
    }
   ],
   "source": [
    "# Create the kallisto index of the simulated Arc transcriptome\n",
    "! kallisto index -i simdata.idx simdata.fasta\n",
    "\n",
    "# Analyze the simulated reads\n",
    "! kallisto quant -i simdata.idx -o . --single -l 150 -s 20 simdata.fastq\n",
    "\n",
    "# Cat the output file so we can see it in our notebook page\n",
    "! cat abundance.tsv\n",
    "\n",
    "# and read in those tpms.\n",
    "tpm_kallisto = get_tpms('abundance.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Does that look more like the correct answer (the $\\nu_i$ we used to create the positive control simulated data), or does it look more like Moriarty's result? Hey, a good reason to plot some graphs with a little matplotlib code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "/* Put everything inside the global mpl namespace */\n",
       "window.mpl = {};\n",
       "\n",
       "\n",
       "mpl.get_websocket_type = function() {\n",
       "    if (typeof(WebSocket) !== 'undefined') {\n",
       "        return WebSocket;\n",
       "    } else if (typeof(MozWebSocket) !== 'undefined') {\n",
       "        return MozWebSocket;\n",
       "    } else {\n",
       "        alert('Your browser does not have WebSocket support. ' +\n",
       "              'Please try Chrome, Safari or Firefox â‰¥ 6. ' +\n",
       "              'Firefox 4 and 5 are also supported but you ' +\n",
       "              'have to enable WebSockets in about:config.');\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure = function(figure_id, websocket, ondownload, parent_element) {\n",
       "    this.id = figure_id;\n",
       "\n",
       "    this.ws = websocket;\n",
       "\n",
       "    this.supports_binary = (this.ws.binaryType != undefined);\n",
       "\n",
       "    if (!this.supports_binary) {\n",
       "        var warnings = document.getElementById(\"mpl-warnings\");\n",
       "        if (warnings) {\n",
       "            warnings.style.display = 'block';\n",
       "            warnings.textContent = (\n",
       "                \"This browser does not support binary websocket messages. \" +\n",
       "                    \"Performance may be slow.\");\n",
       "        }\n",
       "    }\n",
       "\n",
       "    this.imageObj = new Image();\n",
       "\n",
       "    this.context = undefined;\n",
       "    this.message = undefined;\n",
       "    this.canvas = undefined;\n",
       "    this.rubberband_canvas = undefined;\n",
       "    this.rubberband_context = undefined;\n",
       "    this.format_dropdown = undefined;\n",
       "\n",
       "    this.image_mode = 'full';\n",
       "\n",
       "    this.root = $('<div/>');\n",
       "    this._root_extra_style(this.root)\n",
       "    this.root.attr('style', 'display: inline-block');\n",
       "\n",
       "    $(parent_element).append(this.root);\n",
       "\n",
       "    this._init_header(this);\n",
       "    this._init_canvas(this);\n",
       "    this._init_toolbar(this);\n",
       "\n",
       "    var fig = this;\n",
       "\n",
       "    this.waiting = false;\n",
       "\n",
       "    this.ws.onopen =  function () {\n",
       "            fig.send_message(\"supports_binary\", {value: fig.supports_binary});\n",
       "            fig.send_message(\"send_image_mode\", {});\n",
       "            if (mpl.ratio != 1) {\n",
       "                fig.send_message(\"set_dpi_ratio\", {'dpi_ratio': mpl.ratio});\n",
       "            }\n",
       "            fig.send_message(\"refresh\", {});\n",
       "        }\n",
       "\n",
       "    this.imageObj.onload = function() {\n",
       "            if (fig.image_mode == 'full') {\n",
       "                // Full images could contain transparency (where diff images\n",
       "                // almost always do), so we need to clear the canvas so that\n",
       "                // there is no ghosting.\n",
       "                fig.context.clearRect(0, 0, fig.canvas.width, fig.canvas.height);\n",
       "            }\n",
       "            fig.context.drawImage(fig.imageObj, 0, 0);\n",
       "        };\n",
       "\n",
       "    this.imageObj.onunload = function() {\n",
       "        fig.ws.close();\n",
       "    }\n",
       "\n",
       "    this.ws.onmessage = this._make_on_message_function(this);\n",
       "\n",
       "    this.ondownload = ondownload;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_header = function() {\n",
       "    var titlebar = $(\n",
       "        '<div class=\"ui-dialog-titlebar ui-widget-header ui-corner-all ' +\n",
       "        'ui-helper-clearfix\"/>');\n",
       "    var titletext = $(\n",
       "        '<div class=\"ui-dialog-title\" style=\"width: 100%; ' +\n",
       "        'text-align: center; padding: 3px;\"/>');\n",
       "    titlebar.append(titletext)\n",
       "    this.root.append(titlebar);\n",
       "    this.header = titletext[0];\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_canvas = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var canvas_div = $('<div/>');\n",
       "\n",
       "    canvas_div.attr('style', 'position: relative; clear: both; outline: 0');\n",
       "\n",
       "    function canvas_keyboard_event(event) {\n",
       "        return fig.key_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    canvas_div.keydown('key_press', canvas_keyboard_event);\n",
       "    canvas_div.keyup('key_release', canvas_keyboard_event);\n",
       "    this.canvas_div = canvas_div\n",
       "    this._canvas_extra_style(canvas_div)\n",
       "    this.root.append(canvas_div);\n",
       "\n",
       "    var canvas = $('<canvas/>');\n",
       "    canvas.addClass('mpl-canvas');\n",
       "    canvas.attr('style', \"left: 0; top: 0; z-index: 0; outline: 0\")\n",
       "\n",
       "    this.canvas = canvas[0];\n",
       "    this.context = canvas[0].getContext(\"2d\");\n",
       "\n",
       "    var backingStore = this.context.backingStorePixelRatio ||\n",
       "\tthis.context.webkitBackingStorePixelRatio ||\n",
       "\tthis.context.mozBackingStorePixelRatio ||\n",
       "\tthis.context.msBackingStorePixelRatio ||\n",
       "\tthis.context.oBackingStorePixelRatio ||\n",
       "\tthis.context.backingStorePixelRatio || 1;\n",
       "\n",
       "    mpl.ratio = (window.devicePixelRatio || 1) / backingStore;\n",
       "\n",
       "    var rubberband = $('<canvas/>');\n",
       "    rubberband.attr('style', \"position: absolute; left: 0; top: 0; z-index: 1;\")\n",
       "\n",
       "    var pass_mouse_events = true;\n",
       "\n",
       "    canvas_div.resizable({\n",
       "        start: function(event, ui) {\n",
       "            pass_mouse_events = false;\n",
       "        },\n",
       "        resize: function(event, ui) {\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "        stop: function(event, ui) {\n",
       "            pass_mouse_events = true;\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "    });\n",
       "\n",
       "    function mouse_event_fn(event) {\n",
       "        if (pass_mouse_events)\n",
       "            return fig.mouse_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    rubberband.mousedown('button_press', mouse_event_fn);\n",
       "    rubberband.mouseup('button_release', mouse_event_fn);\n",
       "    // Throttle sequential mouse events to 1 every 20ms.\n",
       "    rubberband.mousemove('motion_notify', mouse_event_fn);\n",
       "\n",
       "    rubberband.mouseenter('figure_enter', mouse_event_fn);\n",
       "    rubberband.mouseleave('figure_leave', mouse_event_fn);\n",
       "\n",
       "    canvas_div.on(\"wheel\", function (event) {\n",
       "        event = event.originalEvent;\n",
       "        event['data'] = 'scroll'\n",
       "        if (event.deltaY < 0) {\n",
       "            event.step = 1;\n",
       "        } else {\n",
       "            event.step = -1;\n",
       "        }\n",
       "        mouse_event_fn(event);\n",
       "    });\n",
       "\n",
       "    canvas_div.append(canvas);\n",
       "    canvas_div.append(rubberband);\n",
       "\n",
       "    this.rubberband = rubberband;\n",
       "    this.rubberband_canvas = rubberband[0];\n",
       "    this.rubberband_context = rubberband[0].getContext(\"2d\");\n",
       "    this.rubberband_context.strokeStyle = \"#000000\";\n",
       "\n",
       "    this._resize_canvas = function(width, height) {\n",
       "        // Keep the size of the canvas, canvas container, and rubber band\n",
       "        // canvas in synch.\n",
       "        canvas_div.css('width', width)\n",
       "        canvas_div.css('height', height)\n",
       "\n",
       "        canvas.attr('width', width * mpl.ratio);\n",
       "        canvas.attr('height', height * mpl.ratio);\n",
       "        canvas.attr('style', 'width: ' + width + 'px; height: ' + height + 'px;');\n",
       "\n",
       "        rubberband.attr('width', width);\n",
       "        rubberband.attr('height', height);\n",
       "    }\n",
       "\n",
       "    // Set the figure to an initial 600x600px, this will subsequently be updated\n",
       "    // upon first draw.\n",
       "    this._resize_canvas(600, 600);\n",
       "\n",
       "    // Disable right mouse context menu.\n",
       "    $(this.rubberband_canvas).bind(\"contextmenu\",function(e){\n",
       "        return false;\n",
       "    });\n",
       "\n",
       "    function set_focus () {\n",
       "        canvas.focus();\n",
       "        canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    window.setTimeout(set_focus, 100);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>');\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items) {\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) {\n",
       "            // put a spacer in here.\n",
       "            continue;\n",
       "        }\n",
       "        var button = $('<button/>');\n",
       "        button.addClass('ui-button ui-widget ui-state-default ui-corner-all ' +\n",
       "                        'ui-button-icon-only');\n",
       "        button.attr('role', 'button');\n",
       "        button.attr('aria-disabled', 'false');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "\n",
       "        var icon_img = $('<span/>');\n",
       "        icon_img.addClass('ui-button-icon-primary ui-icon');\n",
       "        icon_img.addClass(image);\n",
       "        icon_img.addClass('ui-corner-all');\n",
       "\n",
       "        var tooltip_span = $('<span/>');\n",
       "        tooltip_span.addClass('ui-button-text');\n",
       "        tooltip_span.html(tooltip);\n",
       "\n",
       "        button.append(icon_img);\n",
       "        button.append(tooltip_span);\n",
       "\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    var fmt_picker_span = $('<span/>');\n",
       "\n",
       "    var fmt_picker = $('<select/>');\n",
       "    fmt_picker.addClass('mpl-toolbar-option ui-widget ui-widget-content');\n",
       "    fmt_picker_span.append(fmt_picker);\n",
       "    nav_element.append(fmt_picker_span);\n",
       "    this.format_dropdown = fmt_picker[0];\n",
       "\n",
       "    for (var ind in mpl.extensions) {\n",
       "        var fmt = mpl.extensions[ind];\n",
       "        var option = $(\n",
       "            '<option/>', {selected: fmt === mpl.default_extension}).html(fmt);\n",
       "        fmt_picker.append(option);\n",
       "    }\n",
       "\n",
       "    // Add hover states to the ui-buttons\n",
       "    $( \".ui-button\" ).hover(\n",
       "        function() { $(this).addClass(\"ui-state-hover\");},\n",
       "        function() { $(this).removeClass(\"ui-state-hover\");}\n",
       "    );\n",
       "\n",
       "    var status_bar = $('<span class=\"mpl-message\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.request_resize = function(x_pixels, y_pixels) {\n",
       "    // Request matplotlib to resize the figure. Matplotlib will then trigger a resize in the client,\n",
       "    // which will in turn request a refresh of the image.\n",
       "    this.send_message('resize', {'width': x_pixels, 'height': y_pixels});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_message = function(type, properties) {\n",
       "    properties['type'] = type;\n",
       "    properties['figure_id'] = this.id;\n",
       "    this.ws.send(JSON.stringify(properties));\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_draw_message = function() {\n",
       "    if (!this.waiting) {\n",
       "        this.waiting = true;\n",
       "        this.ws.send(JSON.stringify({type: \"draw\", figure_id: this.id}));\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    var format_dropdown = fig.format_dropdown;\n",
       "    var format = format_dropdown.options[format_dropdown.selectedIndex].value;\n",
       "    fig.ondownload(fig, format);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_resize = function(fig, msg) {\n",
       "    var size = msg['size'];\n",
       "    if (size[0] != fig.canvas.width || size[1] != fig.canvas.height) {\n",
       "        fig._resize_canvas(size[0], size[1]);\n",
       "        fig.send_message(\"refresh\", {});\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_rubberband = function(fig, msg) {\n",
       "    var x0 = msg['x0'] / mpl.ratio;\n",
       "    var y0 = (fig.canvas.height - msg['y0']) / mpl.ratio;\n",
       "    var x1 = msg['x1'] / mpl.ratio;\n",
       "    var y1 = (fig.canvas.height - msg['y1']) / mpl.ratio;\n",
       "    x0 = Math.floor(x0) + 0.5;\n",
       "    y0 = Math.floor(y0) + 0.5;\n",
       "    x1 = Math.floor(x1) + 0.5;\n",
       "    y1 = Math.floor(y1) + 0.5;\n",
       "    var min_x = Math.min(x0, x1);\n",
       "    var min_y = Math.min(y0, y1);\n",
       "    var width = Math.abs(x1 - x0);\n",
       "    var height = Math.abs(y1 - y0);\n",
       "\n",
       "    fig.rubberband_context.clearRect(\n",
       "        0, 0, fig.canvas.width / mpl.ratio, fig.canvas.height / mpl.ratio);\n",
       "\n",
       "    fig.rubberband_context.strokeRect(min_x, min_y, width, height);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_figure_label = function(fig, msg) {\n",
       "    // Updates the figure title.\n",
       "    fig.header.textContent = msg['label'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_cursor = function(fig, msg) {\n",
       "    var cursor = msg['cursor'];\n",
       "    switch(cursor)\n",
       "    {\n",
       "    case 0:\n",
       "        cursor = 'pointer';\n",
       "        break;\n",
       "    case 1:\n",
       "        cursor = 'default';\n",
       "        break;\n",
       "    case 2:\n",
       "        cursor = 'crosshair';\n",
       "        break;\n",
       "    case 3:\n",
       "        cursor = 'move';\n",
       "        break;\n",
       "    }\n",
       "    fig.rubberband_canvas.style.cursor = cursor;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_message = function(fig, msg) {\n",
       "    fig.message.textContent = msg['message'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_draw = function(fig, msg) {\n",
       "    // Request the server to send over a new figure.\n",
       "    fig.send_draw_message();\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_image_mode = function(fig, msg) {\n",
       "    fig.image_mode = msg['mode'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Called whenever the canvas gets updated.\n",
       "    this.send_message(\"ack\", {});\n",
       "}\n",
       "\n",
       "// A function to construct a web socket function for onmessage handling.\n",
       "// Called in the figure constructor.\n",
       "mpl.figure.prototype._make_on_message_function = function(fig) {\n",
       "    return function socket_on_message(evt) {\n",
       "        if (evt.data instanceof Blob) {\n",
       "            /* FIXME: We get \"Resource interpreted as Image but\n",
       "             * transferred with MIME type text/plain:\" errors on\n",
       "             * Chrome.  But how to set the MIME type?  It doesn't seem\n",
       "             * to be part of the websocket stream */\n",
       "            evt.data.type = \"image/png\";\n",
       "\n",
       "            /* Free the memory for the previous frames */\n",
       "            if (fig.imageObj.src) {\n",
       "                (window.URL || window.webkitURL).revokeObjectURL(\n",
       "                    fig.imageObj.src);\n",
       "            }\n",
       "\n",
       "            fig.imageObj.src = (window.URL || window.webkitURL).createObjectURL(\n",
       "                evt.data);\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "        else if (typeof evt.data === 'string' && evt.data.slice(0, 21) == \"data:image/png;base64\") {\n",
       "            fig.imageObj.src = evt.data;\n",
       "            fig.updated_canvas_event();\n",
       "            fig.waiting = false;\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        var msg = JSON.parse(evt.data);\n",
       "        var msg_type = msg['type'];\n",
       "\n",
       "        // Call the  \"handle_{type}\" callback, which takes\n",
       "        // the figure and JSON message as its only arguments.\n",
       "        try {\n",
       "            var callback = fig[\"handle_\" + msg_type];\n",
       "        } catch (e) {\n",
       "            console.log(\"No handler for the '\" + msg_type + \"' message type: \", msg);\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        if (callback) {\n",
       "            try {\n",
       "                // console.log(\"Handling '\" + msg_type + \"' message: \", msg);\n",
       "                callback(fig, msg);\n",
       "            } catch (e) {\n",
       "                console.log(\"Exception inside the 'handler_\" + msg_type + \"' callback:\", e, e.stack, msg);\n",
       "            }\n",
       "        }\n",
       "    };\n",
       "}\n",
       "\n",
       "// from http://stackoverflow.com/questions/1114465/getting-mouse-location-in-canvas\n",
       "mpl.findpos = function(e) {\n",
       "    //this section is from http://www.quirksmode.org/js/events_properties.html\n",
       "    var targ;\n",
       "    if (!e)\n",
       "        e = window.event;\n",
       "    if (e.target)\n",
       "        targ = e.target;\n",
       "    else if (e.srcElement)\n",
       "        targ = e.srcElement;\n",
       "    if (targ.nodeType == 3) // defeat Safari bug\n",
       "        targ = targ.parentNode;\n",
       "\n",
       "    // jQuery normalizes the pageX and pageY\n",
       "    // pageX,Y are the mouse positions relative to the document\n",
       "    // offset() returns the position of the element relative to the document\n",
       "    var x = e.pageX - $(targ).offset().left;\n",
       "    var y = e.pageY - $(targ).offset().top;\n",
       "\n",
       "    return {\"x\": x, \"y\": y};\n",
       "};\n",
       "\n",
       "/*\n",
       " * return a copy of an object with only non-object keys\n",
       " * we need this to avoid circular references\n",
       " * http://stackoverflow.com/a/24161582/3208463\n",
       " */\n",
       "function simpleKeys (original) {\n",
       "  return Object.keys(original).reduce(function (obj, key) {\n",
       "    if (typeof original[key] !== 'object')\n",
       "        obj[key] = original[key]\n",
       "    return obj;\n",
       "  }, {});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.mouse_event = function(event, name) {\n",
       "    var canvas_pos = mpl.findpos(event)\n",
       "\n",
       "    if (name === 'button_press')\n",
       "    {\n",
       "        this.canvas.focus();\n",
       "        this.canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    var x = canvas_pos.x * mpl.ratio;\n",
       "    var y = canvas_pos.y * mpl.ratio;\n",
       "\n",
       "    this.send_message(name, {x: x, y: y, button: event.button,\n",
       "                             step: event.step,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "\n",
       "    /* This prevents the web browser from automatically changing to\n",
       "     * the text insertion cursor when the button is pressed.  We want\n",
       "     * to control all of the cursor setting manually through the\n",
       "     * 'cursor' event from matplotlib */\n",
       "    event.preventDefault();\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    // Handle any extra behaviour associated with a key event\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.key_event = function(event, name) {\n",
       "\n",
       "    // Prevent repeat events\n",
       "    if (name == 'key_press')\n",
       "    {\n",
       "        if (event.which === this._key)\n",
       "            return;\n",
       "        else\n",
       "            this._key = event.which;\n",
       "    }\n",
       "    if (name == 'key_release')\n",
       "        this._key = null;\n",
       "\n",
       "    var value = '';\n",
       "    if (event.ctrlKey && event.which != 17)\n",
       "        value += \"ctrl+\";\n",
       "    if (event.altKey && event.which != 18)\n",
       "        value += \"alt+\";\n",
       "    if (event.shiftKey && event.which != 16)\n",
       "        value += \"shift+\";\n",
       "\n",
       "    value += 'k';\n",
       "    value += event.which.toString();\n",
       "\n",
       "    this._key_event_extra(event, name);\n",
       "\n",
       "    this.send_message(name, {key: value,\n",
       "                             guiEvent: simpleKeys(event)});\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onclick = function(name) {\n",
       "    if (name == 'download') {\n",
       "        this.handle_save(this, null);\n",
       "    } else {\n",
       "        this.send_message(\"toolbar_button\", {name: name});\n",
       "    }\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onmouseover = function(tooltip) {\n",
       "    this.message.textContent = tooltip;\n",
       "};\n",
       "mpl.toolbar_items = [[\"Home\", \"Reset original view\", \"fa fa-home icon-home\", \"home\"], [\"Back\", \"Back to previous view\", \"fa fa-arrow-left icon-arrow-left\", \"back\"], [\"Forward\", \"Forward to next view\", \"fa fa-arrow-right icon-arrow-right\", \"forward\"], [\"\", \"\", \"\", \"\"], [\"Pan\", \"Pan axes with left mouse, zoom with right\", \"fa fa-arrows icon-move\", \"pan\"], [\"Zoom\", \"Zoom to rectangle\", \"fa fa-square-o icon-check-empty\", \"zoom\"], [\"\", \"\", \"\", \"\"], [\"Download\", \"Download plot\", \"fa fa-floppy-o icon-save\", \"download\"]];\n",
       "\n",
       "mpl.extensions = [\"eps\", \"jpeg\", \"pdf\", \"png\", \"ps\", \"raw\", \"svg\", \"tif\"];\n",
       "\n",
       "mpl.default_extension = \"png\";var comm_websocket_adapter = function(comm) {\n",
       "    // Create a \"websocket\"-like object which calls the given IPython comm\n",
       "    // object with the appropriate methods. Currently this is a non binary\n",
       "    // socket, so there is still some room for performance tuning.\n",
       "    var ws = {};\n",
       "\n",
       "    ws.close = function() {\n",
       "        comm.close()\n",
       "    };\n",
       "    ws.send = function(m) {\n",
       "        //console.log('sending', m);\n",
       "        comm.send(m);\n",
       "    };\n",
       "    // Register the callback with on_msg.\n",
       "    comm.on_msg(function(msg) {\n",
       "        //console.log('receiving', msg['content']['data'], msg);\n",
       "        // Pass the mpl event to the overridden (by mpl) onmessage function.\n",
       "        ws.onmessage(msg['content']['data'])\n",
       "    });\n",
       "    return ws;\n",
       "}\n",
       "\n",
       "mpl.mpl_figure_comm = function(comm, msg) {\n",
       "    // This is the function which gets called when the mpl process\n",
       "    // starts-up an IPython Comm through the \"matplotlib\" channel.\n",
       "\n",
       "    var id = msg.content.data.id;\n",
       "    // Get hold of the div created by the display call when the Comm\n",
       "    // socket was opened in Python.\n",
       "    var element = $(\"#\" + id);\n",
       "    var ws_proxy = comm_websocket_adapter(comm)\n",
       "\n",
       "    function ondownload(figure, format) {\n",
       "        window.open(figure.imageObj.src);\n",
       "    }\n",
       "\n",
       "    var fig = new mpl.figure(id, ws_proxy,\n",
       "                           ondownload,\n",
       "                           element.get(0));\n",
       "\n",
       "    // Call onopen now - mpl needs it, as it is assuming we've passed it a real\n",
       "    // web socket which is closed, not our websocket->open comm proxy.\n",
       "    ws_proxy.onopen();\n",
       "\n",
       "    fig.parent_element = element.get(0);\n",
       "    fig.cell_info = mpl.find_output_cell(\"<div id='\" + id + \"'></div>\");\n",
       "    if (!fig.cell_info) {\n",
       "        console.error(\"Failed to find cell for figure\", id, fig);\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    var output_index = fig.cell_info[2]\n",
       "    var cell = fig.cell_info[0];\n",
       "\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.handle_close = function(fig, msg) {\n",
       "    var width = fig.canvas.width/mpl.ratio\n",
       "    fig.root.unbind('remove')\n",
       "\n",
       "    // Update the output cell to use the data from the current canvas.\n",
       "    fig.push_to_output();\n",
       "    var dataURL = fig.canvas.toDataURL();\n",
       "    // Re-enable the keyboard manager in IPython - without this line, in FF,\n",
       "    // the notebook keyboard shortcuts fail.\n",
       "    IPython.keyboard_manager.enable()\n",
       "    $(fig.parent_element).html('<img src=\"' + dataURL + '\" width=\"' + width + '\">');\n",
       "    fig.close_ws(fig, msg);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.close_ws = function(fig, msg){\n",
       "    fig.send_message('closing', msg);\n",
       "    // fig.ws.close()\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.push_to_output = function(remove_interactive) {\n",
       "    // Turn the data on the canvas into data in the output cell.\n",
       "    var width = this.canvas.width/mpl.ratio\n",
       "    var dataURL = this.canvas.toDataURL();\n",
       "    this.cell_info[1]['text/html'] = '<img src=\"' + dataURL + '\" width=\"' + width + '\">';\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Tell IPython that the notebook contents must change.\n",
       "    IPython.notebook.set_dirty(true);\n",
       "    this.send_message(\"ack\", {});\n",
       "    var fig = this;\n",
       "    // Wait a second, then push the new image to the DOM so\n",
       "    // that it is saved nicely (might be nice to debounce this).\n",
       "    setTimeout(function () { fig.push_to_output() }, 1000);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>');\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items){\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) { continue; };\n",
       "\n",
       "        var button = $('<button class=\"btn btn-default\" href=\"#\" title=\"' + name + '\"><i class=\"fa ' + image + ' fa-lg\"></i></button>');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    // Add the status bar.\n",
       "    var status_bar = $('<span class=\"mpl-message\" style=\"text-align:right; float: right;\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "\n",
       "    // Add the close button to the window.\n",
       "    var buttongrp = $('<div class=\"btn-group inline pull-right\"></div>');\n",
       "    var button = $('<button class=\"btn btn-mini btn-primary\" href=\"#\" title=\"Stop Interaction\"><i class=\"fa fa-power-off icon-remove icon-large\"></i></button>');\n",
       "    button.click(function (evt) { fig.handle_close(fig, {}); } );\n",
       "    button.mouseover('Stop Interaction', toolbar_mouse_event);\n",
       "    buttongrp.append(button);\n",
       "    var titlebar = this.root.find($('.ui-dialog-titlebar'));\n",
       "    titlebar.prepend(buttongrp);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(el){\n",
       "    var fig = this\n",
       "    el.on(\"remove\", function(){\n",
       "\tfig.close_ws(fig, {});\n",
       "    });\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(el){\n",
       "    // this is important to make the div 'focusable\n",
       "    el.attr('tabindex', 0)\n",
       "    // reach out to IPython and tell the keyboard manager to turn it's self\n",
       "    // off when our div gets focus\n",
       "\n",
       "    // location in version 3\n",
       "    if (IPython.notebook.keyboard_manager) {\n",
       "        IPython.notebook.keyboard_manager.register_events(el);\n",
       "    }\n",
       "    else {\n",
       "        // location in version 2\n",
       "        IPython.keyboard_manager.register_events(el);\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    var manager = IPython.notebook.keyboard_manager;\n",
       "    if (!manager)\n",
       "        manager = IPython.keyboard_manager;\n",
       "\n",
       "    // Check for shift+enter\n",
       "    if (event.shiftKey && event.which == 13) {\n",
       "        this.canvas_div.blur();\n",
       "        event.shiftKey = false;\n",
       "        // Send a \"J\" for go to next cell\n",
       "        event.which = 74;\n",
       "        event.keyCode = 74;\n",
       "        manager.command_mode();\n",
       "        manager.handle_keydown(event);\n",
       "    }\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    fig.ondownload(fig, null);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.find_output_cell = function(html_output) {\n",
       "    // Return the cell and output element which can be found *uniquely* in the notebook.\n",
       "    // Note - this is a bit hacky, but it is done because the \"notebook_saving.Notebook\"\n",
       "    // IPython event is triggered only after the cells have been serialised, which for\n",
       "    // our purposes (turning an active figure into a static one), is too late.\n",
       "    var cells = IPython.notebook.get_cells();\n",
       "    var ncells = cells.length;\n",
       "    for (var i=0; i<ncells; i++) {\n",
       "        var cell = cells[i];\n",
       "        if (cell.cell_type === 'code'){\n",
       "            for (var j=0; j<cell.output_area.outputs.length; j++) {\n",
       "                var data = cell.output_area.outputs[j];\n",
       "                if (data.data) {\n",
       "                    // IPython >= 3 moved mimebundle to data attribute of output\n",
       "                    data = data.data;\n",
       "                }\n",
       "                if (data['text/html'] == html_output) {\n",
       "                    return [cell, data, j];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "}\n",
       "\n",
       "// Register the function which deals with the matplotlib target/channel.\n",
       "// The kernel may be null if the page has been refreshed.\n",
       "if (IPython.notebook.kernel != null) {\n",
       "    IPython.notebook.kernel.comm_manager.register_target('matplotlib', mpl.mpl_figure_comm);\n",
       "}\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<img src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAoAAAAHgCAYAAAA10dzkAAAYlklEQVR4Xu3WQQEAAAgCMelf2iA3GzB8sHMECBAgQIAAAQIpgaXSCkuAAAECBAgQIHAGoCcgQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgAABA9APECBAgAABAgRiAgZgrHBxCRAgQIAAAQIGoB8gQIAAAQIECMQEDMBY4eISIECAAAECBAxAP0CAAAECBAgQiAkYgLHCxSVAgAABAgQIGIB+gAABAgQIECAQEzAAY4WLS4AAAQIECBAwAP0AAQIECBAgQCAmYADGCheXAAECBAgQIGAA+gECBAgQIECAQEzAAIwVLi4BAgQIECBAwAD0AwQIECBAgACBmIABGCtcXAIECBAgQICAAegHCBAgQIAAAQIxAQMwVri4BAgQIECAAAED0A8QIECAAAECBGICBmCscHEJECBAgAABAgagHyBAgAABAgQIxAQMwFjh4hIgQIAAAQIEDEA/QIAAAQIECBCICRiAscLFJUCAAAECBAgYgH6AAAECBAgQIBATMABjhYtLgAABAgQIEDAA/QABAgQIECBAICZgAMYKF5cAAQIECBAgYAD6AQIECBAgQIBATMAAjBUuLgECBAgQIEDAAPQDBAgQIECAAIGYgAEYK1xcAgQIECBAgIAB6AcIECBAgAABAjEBAzBWuLgECBAgQIAAAQPQDxAgQIAAAQIEYgIGYKxwcQkQIECAAAECBqAfIECAAAECBAjEBAzAWOHiEiBAgAABAgQMQD9AgAABAgQIEIgJGICxwsUlQIAAAQIECBiAfoAAAQIECBAgEBMwAGOFi0uAAAECBAgQMAD9AAECBAgQIEAgJmAAxgoXlwABAgQIECBgAPoBAgQIECBAgEBMwACMFS4uAQIECBAgQMAA9AMECBAgQIAAgZiAARgrXFwCBAgQIECAgAHoBwgQIECAAAECMQEDMFa4uAQIECBAgACBB+4+AeEFWYvzAAAAAElFTkSuQmCC\" width=\"640\">"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "%matplotlib notebook\n",
    "\n",
    "f, (ax1, ax2) = plt.subplots(1, 2)\n",
    "ax1.loglog(tpm, tpm_kallisto, 'o')\n",
    "expect = [1e3, 1e6]\n",
    "ax1.plot(expect, expect)\n",
    "ax1.set_xlabel('True TPM parameters of the simulation')\n",
    "ax1.set_ylabel('kallisto inferred TPM parameters')\n",
    "\n",
    "ax2.loglog(tpm_moriarty, tpm_kallisto, 'o')\n",
    "ax2.plot(expect, expect)\n",
    "ax2.set_xlabel(\"Moriarty's TPM parameters\")\n",
    "ax2.set_ylabel('kallisto inferred TPM parameters')\n",
    "\n",
    "f.tight_layout()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Huh.** \n",
    "\n",
    "kallisto is overestimating the abundances of the two least abundant Arc transcripts -- thus (possibly) accounting for Moriarty's result."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. 'debug' kallisto\n",
    "\n",
    "My first guess for what's going on was that there's something in `kallisto` that can't deal with the circular permutation of the Arc locus. There's nothing in the `kallisto` paper to suggest it should have a problem with this, and no reason I can think of why it should -- de Bruijn graphs certainly can (and often do) have directed cycles in real data, for example.\n",
    "\n",
    "But... now I'll go back up and rerun the notebook cells but with the `do_no_circle` option on. (This is a hack - again, it'd be better if I'd define a function that runs the simulation!) Note that changing the lengths of transcripts changes their TPM, even though we haven't changed the true nucleotide abundances!\n",
    "\n",
    "Clicky clicky clicky as I rerun the cells down to the point of reloading the abundance.tsv file, and:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'kallisto inferred parameters')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEKCAYAAAD9xUlFAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deZxWc//H8denUUlRKFsklMjOSNY7REIhbrLv3dyy3tYfbutty74LCbf1pjtFiiKJoklRISVLUyRttKhZPr8/zpm5r8YsZ6Y517mumffz8ZjHzHWuc53znjkz87nO+X7P92vujoiICECDpAOIiEjmUFEQEZFSKgoiIlJKRUFEREqpKIiISCkVBRERKbVW0gHWRMuWLb1t27ZJxxARySoTJ0781d1blfdcVheFtm3bkpeXl3QMEZGsYmY/VPScLh+JiEgpFQURESmloiAiIqVUFEREpFTGNDSbWQPgFmA9IM/dn004kohIvRPrmYKZDTCzX8xsapnlh5nZdDObaWZXh4uPAloDBUB+nLlERKR8cV8+GggclrrAzHKAR4DuQEfgRDPrCHQAxrn7ZcD5MecSEclO7jDlNVi1LJbNx1oU3H0MsLDM4k7ATHef5e6rgJcJzhLygUXhOkUVbdPM+phZnpnlzZ8/P47YIiKZadEP8O9e8PrZ8NlzsewiiYbm1sDslMf54bJBQDczewgYU9GL3b2/u+e6e26rVuXekCciUrcUF8MnT8Cje8PsT+Hwu6HT32LZVRINzVbOMnf35cDZ6Q4jIpLR5n8DQy6E2eNhm4Ohx/3Qok1su0uiKOQDW6Q83hyYW50NmFkPoEe7du1qM5eISOYoKoCPHoAP7oSG68DRj8MuvcHKe19de5K4fDQBaG9mW5lZI6A3MKQ6G3D3oe7ep3nz5rEEFBFJ1E+fw5MHwnu3QIfDoe8E2PXE2AsCxHymYGYvAV2AlmaWD9zg7k+bWV9gBJADDHD3aXHmEBHJCgUrgjODjx6Epi3hhH/D9j3SGiHWouDuJ1awfBgwLM59i4hklR/GwZC+sGAm7HYKHHorNFk/7TEy5o7m6lCbgojUGSt/h5E3wYQngwbkUwfDNgcmFicrxz5Sm4KI1AkzRgbdTCc8BXudD+ePS7QgQJaeKYiIZLXlC2HE/8HnL0HLDnD2O7BFp6RTASoKIiLp4w5fvgHDLocVi+CAK+GAy2GtxkknK5WVRUFtCiKSdX7/Gd76B3z9Jmy6K5z6X9hkp6RT/YnaFERE4uQOnz0PD3eCmSPhkJvhnFEZWRAgS88URESywqLvYejFMGs0bLkv9HgQWmb2FQ4VBRGR2lZcBJ/2h1E3g+XAEffCHmdCg8y/OJOVRUFtCiKSsX75OhjALv9TaH8oHHkfNN886VSRZX7ZKofaFEQk4xSugg/6wRP7B3cl93oSTno1qwoCZOmZgohIRpnzWXB2MG8q7HgsHHYnNMvO+V5UFEREaqpgBbx/G4x7GJptDL1fgu0OTzrVGlFREBGpie/HBmcHC2fB7qcHXU2btEg61RrLyqKghmYRScwfv8HIGyBvAKzfFk4bAlv/JelUtUYNzSIiUX3zDjzaGSYOhL37wvkf16mCAFl6piAiklbLFsDwq2HKq9Bqezj+Odg8N+lUsVBREBGpiDtMfR3evjK4bPSXq2H/f8BajZJOFhsVBRGR8vw2NxjAbvow2Gx3OOph2HiHpFPFTkVBRCSVO3z2LLxzPRQVBNNidv47NMhJOllaqCiIiJRYOAuGXATffwht94ceD8CG2ySdKq2ysiioS6qI1KriIhj/GLx3K+Q0DIrBbqdlxQB2tS0rv2N1SRWRWjPvS3j6EHjnWti6C1zwCexxRr0sCJClZwoiImuscBWMvRfG3A1rrwfHPh2MW2SWdLJEqSiISP2TPxGG9IVfvoSd/hoMYNd0w6RTZQQVBRGpP1Yth/f/BeMfhWabwImvQIfDkk6VUaq8aGZmF5vZehZ42sw+M7ND0xFORKTWfDcGHts7GNF099PhgvEqCOWI0pJylrv/BhwKtALOBO6INZWISG35Y0kwT/KzPcAawOlvQo/7YW11VClPlMtHJa0uhwPPuPvnZvW8JUZEMtrgSXPoN2I62/82ltsbPUNLFmP7XARdroFG6yQdL6NFOVOYaGbvEBSFEWa2LlAcb6zKmVkPM+u/ZMmSJGOISAYaPGkOdw36iKuW3cVTje7h1+Km/LXoVga3Ok8FIQJz94qfDM4INie4bDTL3Reb2YZAa3f/Ik0ZK5Sbm+t5eXlJxxCRDDB40hz6Df+aPX4fxY0Nn6UZK3io8BgeL+pJAWvRukUTPrr6oKRjZgQzm+ju5Q7zWunlI3d3Mxvs7nukLFsALKjljCIiNTZ40hweGPQ+N/MUBzeaxKTidlxZ0IcZvnnpOnMXr0gwYfaI0qYw3sz2dPcJsacREamu4mK+eetBhjR4lhyKubngVAYWdaO4zNXxzVo0SShgdolSFA4EzjOz74FlBA3P7u47xxlMRKRKC76FIRdxZeFYxhbvwDWF5zDbN/7Tak0a5nBFtw4JBMw+UYpC99hTiIhUR1EhjH8E3r8Nchpz21oX0P+PffhfZ8n/ad2iCVd068DRu7VOf84sVGVRcPcfzGw/oL27P2NmrYBm8UcTESnHz1ODISrmToIOR8AR99Dx22KaDJrCioKi0tWaNMzh9l47qRhUU5VFwcxuAHKBDsAzQEPg38C+8UYTEUlRuDIYvG7svdBkffjrQOh4NJhx9G7BKv1GTGfu4hVsprODGoty+egYYDfgMwB3nxveqyAikh6zJwRnB/O/hp17w2G3wzobrLbK0bu1VhGoBVGKwqqwa6oDmFnTmDOJiARWLQsmvhn/GKzXGk5+DdofknSqOi1KUXjVzJ4AWpjZucBZwFPxxhKReu/b94Mxixb/AHueC11vgMa6SBG3KA3Nd5vZIcBvBO0K/3T3d2NPVglNxylSh61YHMyCNunfsME2cObbsOU+SaeqNyod5gLAzO5096uqWpYEDXMhUsd89Sa89Q9YNh/2vQj+chU01E1nta2yYS6iDIhX3gU83bsgIrVn6S/w6unwysnQtBWcOwq63qiCkIAKLx+Z2fnA34GtzSx18Lt1gY/iDiYi9YA7fPEKDL86aFQ+6HrY92LIaZh0snqrsjaFF4G3gduBq1OW/+7uC2NNJSJ13+LZ8OalMPNd2GIv6PkQtNJQFEmrsCi4+xJgCXBimTuaW5rZVu7+XdpSikjdUVwMeU/DyBuDM4XudwW9ixpEuZotcavJHc2N0B3NIlITv86AIRfCj+Ng6wOhxwOw/pZJp5IUuqNZROJXVAgfPwij74CGa8NRj8KuJ4Fm9s04uqNZROL10xfBEBU/fQ7b94DD74F1/zy8tWSGmt7R/GS8sUQk6xX8AWPugrH3wzobwvHPQcejkk4lVcjKO5pFJMP9+ElwdvDrN7DryXDorX8awE4yU5QzBdz9XTP7pGR9M9tA3VJF5E9WLoVRN8On/aH5FnDKIGh3cNKppBqi9D76G3AzsAIoJpyOE9g63mgiklVmjoKhl8CS2dCpDxz8T2is+biyTZQzhcuBHdz917jDiEgWWr4Q3rkOJr8AG7aHs4ZDm85Jp5IailIUvgWWxx1ERLLQl2/AW5fD8gWw/z/ggCuDLqeStaIUhWuAj8M2hZUlC939othSiUhm+30eDLscvhoCm+wMp7wOm+6cdCqpBVGKwhPAe8AUgjYFEamv3GHyizDi/6BgRTCS6d4XQk6kPiuSBaIcyUJ3vyzuIGbWBbgFmAa87O6j496niFTDoh+CmdBmvQ9t9g4GsGvZPulUUsuijED1vpn1MbNNzWyDko8oGzezAWb2i5lNLbP8MDObbmYzzaxkBFYHlgJrA/nV+i5EJD7FxfDJE/Do3pA/AQ6/G84YpoJQR0U5Uzgp/HxNyrKoXVIHAg8Dz5UsMLMc4BGCyXvygQlmNgT40N0/MLONgXuBkyNsX0TiNH96MIDd7E+gXVc48j5o0SbpVBKjKHc0b1XTjbv7GDNrW2ZxJ2Cmu88CMLOXgaPc/cvw+UVA44q2aWZ9gD4Abdrol1OkNg2eNId+I6bzy+LfubzZCM4pfpWcxs3gmCdg5xM0gF09EKl1yMx2BDoSXNoBwN2fq/gVlWoNzE55nA/sZWa9gG5AC4Kzi3K5e3+gPwRzNNcwg4ikGDxpDjcNncai5QXsYN/Rv1F/dij8gbeL94ZD7qT7LrskHVHSJOp8Cl0IisIwgvmZx5JySaiaynur4e4+CBhUw22KSA1dN3gKL4z/kUas4sq1BtEn500Wsh59Vl3KO8V70vqDBXTfO+mUki5RzhSOA3YBJrn7meE1/6fWYJ/5wBYpjzcH5lZnA2bWA+jRrl27NYghIoMnzeGF8T+Sa19zR8Mn2abBT7xc2IXbCk/iN4IhKuYuXpFwSkmnKL2PVrh7MVBoZusBv7Bm4x5NANqb2VZm1gjoDQypzgbcfai792nevPkaxBCp3wZPmsM/Xx3PTWs9w38a30wjCjl51TVcXdintCAAbNaiSYIpJd2inCnkmVkLgjkUJhJ0G/00ysbN7CWCS08tzSwfuMHdnzazvsAIIAcY4O7TahJeRGpm8KQ5DBv0HG83epJNWcjThd25u/CvrGD1ISqaNMzhim4dEkopSai0KJiZAbe7+2LgcTMbDqzn7l9E2bi7n1jB8mEE7RMikm7LF9J46N/pnzOaGcWtOa7gBj7zbf+0WosmDbmx5w4cvVvrBEJKUiotCuE0nIOBPcLH36cjVFXUpiBSA+7w5WAYdgVdixbyQNExPFJ4NKtouNpqBpzcuQ23Hr1TMjklUVHaFMab2Z6xJ6kGtSmIVNNvP8Erp8B/zoD1WnN247u5r/CvfyoIOWbcd8KuKgj1WJQ2hQOBv5nZD8Aywkl23F1DIopkOneY9DyMuA6KVsIhN0PnC+j1xTwmDJrCioKi0lWbNMzh9l476XJRPRelKHSPPYWI1L6F3wUD2H33AWy5bzCA3YbbAJT+4+83YjpzF69gsxZNuKJbBxUEiTTMxQ8AZrYRkBGzZ6hNQaQSxUXBAHbv3QKWA0fcC3ucCQ1Wv1p89G6tVQTkT6psUzCznmY2A/gO+AD4Hng75lyVUpuCSAV++RoGdIMR10Db/eGC8bDn2X8qCCIViXL56BagMzDS3XczswOBcruaikhCClfBR/fDmH7QqBn0egp2Ok4D2Em1RSkKBe6+wMwamFkDd3/fzO6MPZmIRDPns2B463lTYcdjoftd0LRl0qkkS0UpCovNrBnwIfCCmf0CFMYbq3JqUxABVi2H0bfDuIeh2cbQ+yXY7vCkU0mWi3Kh8ShgBXAJMBz4FugRZ6iqqE1B6r3vx7L0gb3g4wd5qeAvHLqqH4NXaHhrWXNReh8tM7NNCCbHWQiMcPcFsScTkT/74zcYeQPkDWChb8y5BdcyrngHWALXDJoCoB5Fskai9D46h2AAvF4Ew2iPN7Oz4g4mImV8MwIe7QwTB/JizlEcuvKOoCCEVhQU0W/E9AQDSl0QpU3hCmC3krMDM9sQ+BgYEGcwEQkt+xWGXw1T/gMbdYTjn+fah3+mvGkHNfeBrKkobQr5wO8pj39n9ek0087MephZ/yVLliQZQyRe7jDlNXikE0wbDF2ugT4fwOZ7VDjHgeY+kDUVpSjMAT4xsxvDqTnHAzPN7DIzuyzeeOVTQ7PUeb/NhZdOhNfPhvXbwt/GQJerYa1GAFzRrQNNGuas9hLNfSC1Icrlo2/DjxJvhJ/Xrf04IvWcO3z2LLxzPRQVwKH/gs7nQ4PVC4DGLpK4mHt5VyazQ25urufl5SUdQ6R2LJwFQy6C7z8Mhqjo+SBssCYz34qUz8wmuntuec9FOVMQkTgVF8H4x+C9WyGnIfR4AHY/XUNUSCJUFESSNO9LGNIX5kyEbbvDkffCepslnUrqMRUFkSQUroIP7wk+1m4Oxw2AHXrp7EASV2FRMLOHoNyu0AC4+0WxJIpAYx9JVsufCG9cAPO/gp2Oh8PugKYbJp1KBKi8S2oeMJFgYp3dgRnhx65AUSWvi526pEpWWrUcRlwLT3eFlb/BSa/CsU+qIEhGqfBMwd2fBTCzM4AD3b0gfPw48E5a0onUFd+NCYa3XvQ95J4FXW+CtddLOpXIn0RpU9iM4J6EheHjZuEyEanKH0uCew4+ezboXnrGW9B2v6RTiVQoSlG4A5hkZu+Hj/8C3BhbIpG64uth8NZlsHQe7HtxMExFQw1DIZktytDZz5jZ28Be4aKr3f3neGOJZLGl8+HtK2HaINhoB+j9IrTePelUIpFUWRTMzICuwNbufrOZtTGzTu7+afzxRLKIezCS6dtXwaqlcOB1wRlCOF6RSDaIcvnoUaAYOAi4mWCU1NeBPWPMJZKIwZPm1Gw8oSX58OZlMGMEbL4n9HwYNtou/sAitSxKUdjL3Xc3s0kA7r7IzBJ966P7FCQOgyfN4ZpBU1hREPS4nrN4RdWzmRUXw8Rn4N0bwIuCew469fnTAHYi2SLK0NkFZpZDeCObmbUiOHNIjO5TkDj0GzG9tCCUqHQ2swXfwrNHBo3Jm+8Bfx9X7oimItkkypnCg8B/gY3M7F8EU3JeF2sqkQRUNGvZn5YXFcK4h2H07ZDTOLhUtNspGqJC6oQovY9eMLOJwMGAAUe7+1exJxNJs81aNGFOOYVhtdnMfp4Cb/SFnybDdkfC4XfDepumMaVIvCotCmbWAPjC3XcEvk5PJMlUNW6EzRJXdOuwWpsCpMxmVrgSxvSDsfdBk/XhrwOh49E6O5A6p9Ki4O7FZva5mbVx9x/TFUoyT40aYbNMhbOZtZwDjx8Dv06HXU6EbrfBOhsknFYkHlHaFDYFppnZp8CykoXu3jO2VJJxKmuErStFAYLCUPr9rFoGo26BNx6H5pvDya9D+67JBhSJWZSicFPsKSTjRW6ErSu+fR+GXgSLf4Q9z4WuN0BjTUsudV9VbQo5wPXurrdH9VykRti6YMUieOc6mPRv2LAdnPk2bLlP0qlE0qbS+xTcvQhYbma6IaCeu6JbB5o0XL3/fWkjbF3x1VB4ZC+Y/BLsdymc95EKgtQ7US4f/QFMMbN3Wb1NIbGZ1yT9KmyErQvtCUt/gWFXwJeDYZOdgslvNts16VQiiYhSFN4KPzKGhrlIxmqNsBFkfBdWd/j8ZRh+NRQsh4OuDwawy2mYdDKRxJh7hdMw/28lsyZAG3ev4H7/ZOTm5npeXl7SMaQcZbuwQnC56fZeO2VGYVg8G968BGaOhC32Cu5KbrVt0qlE0sLMJrp7bnnPVTn2UfiufDIwPHy8q5kNqd2IUtdUexyhdCkuhk+fhEc7ww/joHs/OHO4CoJIKMrloxuBTsBoAHefbGZbxZhJ6oCM7ML664xgnuQfx8E2B8GR98P6WyaXRyQDRSkKhe6+xFa/nb/qa05Sr2VUF9aiAvj4IRh9RzAd5tGPBXcma4gKkT+JMnT2VDM7Ccgxs/Zm9hDwccy5JMtlTBfWnz6HJw+CUTfBtt3ggk9h15NUEEQqEOVM4ULgWmAl8CIwArg1zlCS/RLvwlrwB4y5C8beD+tsCMc/Bx2PSs++RbJYhb2PzOx5dz/VzC529wfSnCsS9T6Scv04PhjeesEM2PUUOPQWDWAnkqKy3keVnSnsYWZbAmeZ2XMEcymUcveFtZhRZM2tXAqjboZP+0PzLeCUQdDu4KRTiWSVyorC4wTdULcGJrJ6UfBwuUhmmDkShl4CS/Jhr78FN6I1bpZ0KpGsU2FRcPcHgQfN7DF3Pz+NmUSiW74QRlwLn78ILbeFs4ZDm85JpxLJWlGm4zw/HC1149T1NemOJO7LN+Cty2H5Atj/cjjgCmi4dtKpRLJalUXBzPoS3MA2DygOFzuwc3yxRCrx+88w7PJgVNNNd4FTXodN9esoUhuidEm9BOjg7gviDiNSKXeY/CKMuCboctr1Rtj7QsiJ8mssIlFE+WuaDSyJO4hIpRb9AEMvhlnvQ5t9oOdD0FKj5IrUtihFYRYw2szeIriBDQB3vze2VCIliouCAexG3RzchXz43ZB7NjSIcjO+iFRXlKLwY/jRKPwQSY/504MB7GZ/Au26BgPYtdgi6VQidVqU3kc3pSMIgJk1BcYAN7j7m+nar2SYogL46H744C5o1BSOeQJ2PkHjFYmkQYVFwczud/dLzGwo5YyK6u49q9q4mQ0AjgR+cfcdU5YfBjwA5ABPufsd4VNXAa9W71uQOmXu5GCIinlTYIdjoPtd0GyjpFOJ1BuVnSk8H36+ew22PxB4GHiuZEF4z8MjwCFAPjAhnLRnM+BLQB3N66OCFcHQ1h8/BE1bwQkvwPZHJp1KpN6p7I7mieHnD2q6cXcfY2ZtyyzuBMx091kAZvYycBTQDGgKdARWmNkwdy8u81rMrA/QB6BNmzY1jSaZ5PuPgraDhd/CbqfCobdCkxZJpxKpl5Lo4N2aoJtriXxgL3fvC2BmZwC/llcQANy9P9AfglFS440qsfrjt2CegwlPQYst4bQ3YOsuSacSqdeSKArltRaW/nN394HpiyKJmfFuMIDdb3Og89/hoOuCRmURSVTkomBm6wLu7kvXcJ/5QGq/ws2BudXZgJn1AHq0a6ebl7LO8oUw/Br44mVotR2c/S5ssWfSqUQkVOUdQGa2k5lNAqYCX5rZRDPbsarXVWIC0N7MtjKzRkBvYEh1NuDuQ929T/PmzdcghqSVO0wdBA/vCVNfg79cBX8bo4IgkmGinCk8AVzm7u8DmFkXgmv6+1T1QjN7CegCtDSzfIL7D54OB9kbQdAldYC7T6tZfMkKv/0Eb/0Dpr8Fm+0GPd+ATdbkfYWIxCVKUWhaUhAA3H10eJNZldz9xAqWDwOGRYsoWcsdJj0PI66DopVwyC1B+4EGsBPJWJHGPjKz6/nffQunAN/FF6lqalPIAgu/g6EXwXdjYMv9oOeDsOE2SacSkSpEGVXsLKAVMCj8aAmcEWOmKqlNIYMVF8G4R+GxfWDOJDjyPjh9qAqCSJaIcqbQ1d0vSl1gZn8F/hNPJMlav3wVDFExJw/adwsKQvPWSacSkWqIcqZwTcRlaWNmPcys/5IlmuYhIxSuCgave3x/WPQdHPs0nPSKCoJIFqpsQLzuwOFAazN7MOWp9YDCuINVxt2HAkNzc3PPTTKHAHMmwhsXwi/TYMfjoPud0LRl0qlEpIYqu3w0F8gDegITU5b/DlwaZyjJAquWw+jbYNwj0GwTOPFl6NA96VQisoYqGxDvc+BzM3vR3QsAzGx9YAt3X5SugJKBvvsw6Fm0cBbscQYccjOsrUZ/kbogSkPzu2bWM1x3MjDfzD5w98vijSYZ548l8O4NMPEZWH+roFfRVgcknUpEalGUotDc3X8zs3OAZ9z9BjP7Iu5gldF9Cgn4ZkQwgN3Sn2HvvnDgtdBonaRTiUgti9L7aC0z2xQ4HsiIKTJ1n0IaLfsVXj8HXjw+mOPg7JHQ7V8qCCJ1VJQzhZsJxika6+4TzGxrYEa8sSRx7jD1dXj7ymDegy7/B/tdCms1SjqZiMSoyqLg7v8h5Ua1cMa0Y+MMJQlbMgfeugy+GQ6t94CeD8PGHZNOJSJpUNl9Cle6+11m9hApk+CUKHuXs9QBxcXw2bPw7j+hqAC63QZ7nQcNcpJOJiJpUtmZwlfh57x0BKkONTTHYMG3MPRi+P5DaLt/MIDdBlsnnUpE0szcs3ea49zcXM/Ly7ialV2Ki2D8o/DevyCnIRx6K+x+Glh5s6aKSF1gZhPdPbe85yq7fDSUci4blXD3nrWQTZI0b1owgN3cz6DD4XDEPbDeZkmnEpEEVXb56O60pZD0KlwJH94TfKzdAo4bADv00tmBiFQ6zMUH6QwiaZKfF5wdzP8Kdj4But0OTTdMOpWIZIjKLh9NofLLRzvHkkjisWpZ0G4w/tHgEtFJr8K23ZJOJSIZprLLR0emLUU1qfdRNc36IBjAbtH3kHs2dL0R1l4v4VAikonU+6guW7EY3r0ePnsONtgm6Gbadr+kU4lIwmrU+yjlxZ2Bh4DtgUZADrDM3fVWM5N9PSy4K3npPNj3YuhyDTRsknQqEclwUcY+ehjoTTDURS5wGqDrNplq6fxgvKJpg2DjHaH3i9B696RTiUiWiFIUcPeZZpbj7kXAM2b2ccy5pLrc4YtXYfhVQaPygdfBfpcEN6SJiEQUpSgsN7NGwGQzuwv4CWgabyypliX58OalMOMd2HzPYAC7jbZLOpWIZKEoReFUgnkX+hLMzbwFGiU1MxQXw8QB8O6N4EVw2J3Q6VwNYCciNRalKLR09x+AP4CboLRL6Mw4g0kVFnwLQy6EHz6CrbtAjwdg/bYJhxKRbBdl5rUnzWynkgdm1hu4Lr5IUqmiQhh7Pzy2D8ybCkc9AqcOVkEQkVoR5UzhOOA1MzsZ2I+g99GhsaaqQr29ee3nKfDGBfDT57DdkcEAdutuknQqEalDIt28ZmbbAoOB2cDR7r4i7mBR1Jub1wpXwph+MPY+aLI+HH43dDxKA9iJSI3UdOjssmMfbUBw49onZqaxj9Jl9qfBAHa/ToddTgxmQ1tng6RTiUgdlZVjH9ULK5fCe7fCJ49D883h5NehfdekU4lIHVfZ0Nk/pDOIpPj2vWBqzMU/Qqc+cPA/ofG6SacSkXog0h3NkiYrFsGI62Dyv2HD9nDmcNhy76RTiUg9oqKQKb4aCm/9A5b9CvtdBn+5ChqunXQqEalnVBSS9vs8ePsK+PIN2GSnYPKbzXZNOpWI1FMqCklxh89fhuFXQ8GKoN1gn4s0gJ2IJEpFIQmLf4Shl8C3o2CLztDzIWi1bdKpRERUFNKquBjynoaRNwZnCt37wZ7nQIMoo42IiMQvK4tCVg5z8euM4Ca02eNhm4Ohx/3Qok3SqUREVpOVb1Hdfai792nevHnSUapWVAAf3gOP7Qvzv4ajH4NTXldBEJGMlJVnClnjp8+Ds4OfvwjGKureD9bdOOlUIiIVUlGIQ8Ef8MGd8NEDsM6GcPzz0LFn0qlERKqkolDbfhwfnB0smAG7ngLdbg1GNhURyQIqCplJZm4AAAv+SURBVLVl5e8w6mb49ElosQWc+l/Y5qCkU4mIVIuKQm2YOTK472BJPuz1NzjoemjcLOlUIiLVpqKwJpYvhBH/B5+/BC23hbNGQJu9kk4lIlJjKgo1NW0wDLs8GNl0/8vhgCs0gJ2IZD0Vher6/eegGHw1FDbdBU4ZBJtqEjoRqRtUFKJyh8kvBJeLCldC15tg776Qox+hiNQd+o8WxaLvg5nQZo2GNvsEA9i1zKIhNkREIlJRqExxUdDFdNRNYA3giHtgj7M0gJ2I1FkqChWZPz24CS3/U2h3CBx5X3D/gYhIHaaiUFZRAXx0P3xwFzRqCsf0h52PB7Okk4mIxE5FIdXcScHZwbypsEMv6H4XNGuVdCoRkbRRUYBgOszRt8PHD0PTVtD7RdjuiKRTiYikXcYUBTPbHrgYaAmMcvfH0rLj7z+CIRfCwm9h99PgkFugSYu07FpEJNPE2o3GzAaY2S9mNrXM8sPMbLqZzTSzqwHc/St3Pw84HsiNMxcAf/wGb14GAw+H4kI47Y2gq6kKgojUY3H3rRwIHJa6wMxygEeA7kBH4EQz6xg+1xMYC4yKNdXMUfDo3pA3ADpfAH8fB1t3iXWXIiLZINai4O5jgIVlFncCZrr7LHdfBbwMHBWuP8Td9wFOrmibZtbHzPLMLG/+/Pk1C/bb3GAU07PfhcNuC3oZiYhIIm0KrYHZKY/zgb3MrAvQC2gMDKvoxe7eH+gPkJub6zVKsNspsPMJsFajGr1cRKSuSqIolNfh3919NDA6PQlMBUFEpBxJjNeQD6TeGrw5MLc6GzCzHmbWf8mSJbUaTESkvkuiKEwA2pvZVmbWCOgNDKnOBtx9qLv3ad68eSwBRUTqq7i7pL4EjAM6mFm+mZ3t7oVAX2AE8BXwqrtPizOHiIhEE2ubgrufWMHyYVTSmCwiIsnIyjGg1aYgIhKPrCwKalMQEYlHVhYFERGJh7nX7P6vTGBm84EfyixuDpR3Xans8pbArzFFq0pFGePeTtT1q1qvsuej/vwrWpbUcUnqmFTnNTU9Lmu6XH8rNV8vU/9WtnT38ucFcPc69QH0j7IcyMu0jHFvJ+r6Va1X2fNRf/6VLEvkuCR1TNJxXNZ0uf5Wav+YVPe4pPNvpS5ePhpazeVJqK0s1d1O1PWrWq+y56vz89cxqd5ranpcamt5EvS3Em0/tSarLx+tCTPLc/f4h+iWatFxyTw6JpkpruNSF88UouqfdAApl45L5tExyUyxHJd6e6YgIiJ/Vp/PFEREpAwVBRERKaWiICIipVQUQma2vZk9bmavmdn5SeeRgJk1NbOJZnZk0lkkYGZdzOzD8O+lS9J5BMysgZn9y8weMrPT12RbdboomNkAM/vFzKaWWX6YmU03s5lmdjWAu3/l7ucBxwPqfheT6hyT0FXAq+lNWf9U87g4sBRYm2DSLIlBNY/JUQRTHRewhsekThcFYCBwWOoCM8sBHgG6Ax2BE82sY/hcT2AsMCq9MeuVgUQ8JmbWFfgSmJfukPXQQKL/rXzo7t0JCvZNac5Znwwk+jHpAIxz98uANbrSUaeLgruPARaWWdwJmOnus9x9FfAyQZXF3Ye4+z7AyelNWn9U85gcCHQGTgLONbM6/fuapOocF3cvDp9fBDROY8x6pZp/K/kExwOgaE32G+skOxmqNTA75XE+sFd4bbQXwS+5JgBKr3KPibv3BTCzM4BfU/4ZSXpU9LfSC+gGtAAeTiJYPVbuMQEeAB4ys/2BMWuyg/pYFKycZe7uo4HR6Y0ioXKPSekX7gPTF0VSVPS3MggYlO4wAlR8TJYDZ9fGDurj6Xg+sEXK482BuQllkYCOSWbScck8sR+T+lgUJgDtzWwrM2sE9AaGJJypvtMxyUw6Lpkn9mNSp4uCmb0EjAM6mFm+mZ3t7oVAX2AE8BXwqrtPSzJnfaJjkpl0XDJPUsdEA+KJiEipOn2mICIi1aOiICIipVQURESklIqCiIiUUlEQEZFSKgoiIlJKRUEwsw3NbHL48bOZzUl53CjpfHELx6K/uuo1a3WfTczsvfBnfFyZ584ys01SHuebWYta3v8IM1u3FrazlpktrmKdDczsvJTHW5jZK2u6b4mH7lOQ1ZjZjcBSd7+7zHIj+H1JbFA6M8tx9zUaAbKC7a5FMOBetf7xrkkeM9sPuMndDy7nubFAX3efHD7OB3Z090r/+SYhys/OzNoBr7n7rulLJjWlMwWpkJm1M7OpZvY48BmwReq7QjPrbWZPhV9vbGaDzCzPzD41s87lbO8cM/tv+C51upldl/LcUAtmWJtmZueEy9Yys8VmdquZfQp0MrObzGxCSa6wWGFmY83sXgtmBPvSzHLDfc0IC13Jfk4P8002s0fD4bjvANYNlz1X0XoV5OkX7u8LM7uznO+5pZkNCZ//2Mx2NLPNCMbKzw233zZl/ROAXYFXypypXWJmk8LtbBuu28zMBoY5J5lZj3L23zr82UwOf2b7hMvzzaxFyjEeEP7snzOzbmHWb8wsN1z/VjO7JGW7X5vZ5mX2tV549vNZmLNktrw7CO7KnWxmd4T7LCl4TczsWTObEr7ugJTfldfC35UZZnZ72e9NYuLu+tBH6QdwI3B5+HU7oBjYM3y8FrA4Zd3ewFPh168AncOv2wJTy9n2OcAcYH2gKcEEOruGz20Qfl4nXL5+uD8HeqVso2Q9A14CuoePxwL/Cr/+B8HAYRsTzA42l2CY5x2BwcBa4Xr9CeZqKPt9VbZeaZ5w+9P43xl3i3K+58eAa8OvDwXywq+7AoMrOAZjS34u4eN84Pzw64uAx8Ov7wJ6h1+vD3wDrF1mW1cBV4Vf5wDNUrbZIjzGBQQTtjQAJgNPhuscS/AOH+BW4JKU7X5NMBhb6c8OaAisG369ETAj5fdocsprSx+H+Ur2twPwA9Ao/F2ZAawLNCEYLnqzpP8+6sNHfRw6W6rnW3efEGG9rgTvBkser29mTdx9RZn1Rrj7IgAzGwzsR/CP6FILZr6D4J/NNuHyVcB/U15/sJldQfDPviUwEXg7fK5kYLApwBR3nxfu5/twm12BPYG8MGfJP5vyvpeK1kvNs5CgaD5pZm8Bb5azrf2AIwDc/Z3wnX3TctarSslQ1ROBw8OvDwW62//aQ9YG2hAUhxITgCfMbG2CIvR5Odue6e5fApjZl8DIcPkU4JpqZDTgTgsujRUTnFm2rOI1+wH9ANx9mpnNJSgaACPd/fcw19fh96ZRWmOmoiBVWZbydTGrj+e+dsrXBnTyYDaoypRtxHILpt08gOBMY4UF19RLtr3Cw7eRZrYOwaQuu7v7HDO7tUyGlSk5V6YsLyb4XTdggLtfnxrAguviqy2qZL3SPO5eEF5eOYTgrOl8gn/UZbdV2eOoSr6fIv73d2vA0e7+bUUvcvf3LJhA6gjgBTO73d1fqGDbsPrPruTnBlDI6pebU3/uJU4DmhMcn0IL2kLKWy9VZT+P1Fyp37fESG0KEpkHjcyLzKx9eC3+mJSnRwIXlDwws4oaFQ8Nr2WvQzCN4EcE/0gWhgVhB4J36eVpQvCP6lcLes4cW81vYSRwfMm7Vwt6XbXxYOTJ1OJQ7nplNxZmWM/d3wQuBXYrZ59jCKd3DYtfvrsvK2e9VL8TXDapygiCy0klef60fzPbEvjZ3fsTtGOUlzGK74E9wm12YvUx/Us0B34JC8IhBLOEQeXfT+rPZ3tgU2BmDTNKLVDlleq6ChgO/Ehw7b9kjt4LgMfM7EyC36v3SSkSKcYCLxJcHnre3SeHlwb6mNnnBNeqPylvx+6+wMyeBaYSXHsud72KuPsUM7sJGBkWtQLgvPB7eRr4wszy3P20CtYre+miOTDIzBoTvMG6rJzd/hN4xsy+AJYCZ0aI+gzwlJmtIJiTtyI3Afeb2ZRw/zMJ5xtPcTBwmZkVhPs/JcL+y/Mf4BQzmwR8CswqZ53ngaFmlkfQMWEGgLvPs6ADwhTgLeCplNc8RHB5awrBz/k0d1+VchlS0kxdUiVtLOhVtKO7X1LlyiKSCF0+EhGRUjpTEBGRUjpTEBGRUioKIiJSSkVBRERKqSiIiEgpFQURESmloiAiIqX+H6a20iNcfpqOAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "f, ax3 = plt.subplots()\n",
    "ax3.loglog(tpm, tpm_kallisto, 'o')\n",
    "ax3.plot(expect, expect)\n",
    "ax3.set_xlabel('True parameters of the simulation')\n",
    "ax3.set_ylabel('kallisto inferred parameters')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interesting. Now it's getting Arc1 right. But it's still messing up on Arc6.\n",
    "\n",
    "Another 'debugging' control we can run is a drastic one - let's remove _all_ the overlaps between Arc transcripts, and make sure that `kallisto` works in the trivial case. \n",
    "\n",
    "Again, hacking - go back up and set the `do_no_overlap` option, and rerun the notebook, including the graphs above. Now we see `kallisto` nail it, as it should.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions\n",
    "\n",
    "A little puzzling, actually! De-circularizing the Arc locus fixed Arc1 but not Arc6, so it's not solely a matter of circularization. Removing all overlaps (so all reads uniquely map) fixes the problem completely, but of course it does; it makes the problem trivial, and the whole point of kallisto is to infer transcript abundances even when transcripts overlap.\n",
    "\n",
    "We've run this problem every year that MCB112 has been taught and it's still not clear what's going on. And new versions of kallisto continue to show the same behavior. The best solution that I've seen was contributed by Will Mallard, one of the TFs in 2017. You can [download his pset solution .ipynb file from 2017](http://mcb112.org/w02/w02-answers-WM.ipynb). Will's experiments show that for _low abundance_ transcripts, kallisto needs _at least one piece of that transcript to be unique_ (i.e. reads map uniquely to it). De-circularization, in Will's view, is a red herring; what also happens when we decircularize is that the A segment becomes unique to Arc1, and kallisto accurately quantitates Arc1. When he did simulations where he removes overlaps from the start of Arc6, kallisto accurately quantitates Arc6. \n",
    "\n",
    "Again, there's no reason (mathematically) why kallisto should have a problem with the overlaps, even though our Arc locus is a pretty pathological case... and it's puzzling that it only has a problem with quantitating low-abundance transcripts. But Will's dissection is the best so far."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
